### PR TITLE
Add Copy as cURL context menu to network tab

### PR DESCRIPTION
**Pull request by vikneshwar**
This PR fixes the #8789 
- Adds the copy as cURL context menu in the network tab
  - However, there is one issue for which we need to decide how we go about handling it
     - When you right-click and choose `Copy as cURL`  for a request without actually selecting and opening the request first, the request body won't be copied in the cURL, this is because we download the body for the particular request only when you click and select the request.So no request body is available to be copied. 
     - But once it is downloaded(by clicking on the request and trying it after a few seconds) and we try to copy, the body is copied exactly as expected
     - This also causes one more issue which we might have left unnoticed. For the first time, as soon as you select a request and navigate to Request or Response body tab, you won't be able to see any information because the socket call is yet to be completed
     - I can think of 2 solutions here (but looking for better suggestions from others):
        - Download the request and response body for all the requests when we render the network panel. But that will result in significant calls, which is useless (most users won't check all the requests). This also fixes both the above-mentioned issues
        - When the user clicks `copy as cURL`, we make the socket call for the particular request to fetch the request and response body, show the loading state until the call completes, and copy to the clipboard. Similarly, when the user clicks on a request and navigates to the request tab quickly, we can show the loading state 
- Removes the root object in the request body json view. 
  - Currently, when you view the request payload by clicking on the request tab, the body is enclosed inside the `root` object, which is not expected. This PR fixes that too. 

Replay: https://app.replay.io/recording/overboard--999b4bf9-463e-4ca5-9336-7b585e28ad5b

### CODE REVIEW

**Comment by vikneshwar** [comment]
A couple of unit test fails when i run `npm run test`, but those are unrelated to the changes, I tried to run test in main branch and it still fails.

**Comment by vikneshwar** [comment]
@jasonLaster can you review this ?

**Comment by bvaughn** [comment]
> When the user clicks copy as cURL, we make the socket call for the particular request to fetch the request and response body, show the loading state until the call completes, and copy to the clipboard. Similarly, when the user clicks on a request and navigates to the request tab quickly, we can show the loading state

This (fetch on demand, if needed) seems like the right way to go.

**Comment by bvaughn** [comment]
Delta is suggesting this PR also regresses another recently added context menu. For example:

Before:
image

After:
image

I think this is just indicating that you need to rebase this branch on `main` though

**Comment by vikneshwar** [comment]
@bvaughn Thanks for your suggestion. I have merged the latest changes and also implemented the suggested solution. Let me know how it looks now. 

Changes done as part of the previous commit:
 - Create a hook to handle copy to clipboard state.
- Show Copying status until we get the request body from the server.
- Copy the body and show Copied status as soon as we get a response.
- For all other cases (i.e with no request body or body is already available) copy immediately and show copied status.

File: src/ui/components/NetworkMonitor/RequestTable.tsx [hunk 1]
```diff
@@ -36,12 +37,14 @@ const RequestTable = ({
 
   const loadedRegions = useAppSelector(getLoadedRegions);
 
+  const [showCopied, setShowCopied] = useState(false);
```
Comment (bvaughn) [inline src/ui/components/NetworkMonitor/RequestTable.tsx:40]: Why is this state here? Doesn't look like it's being read or written.

File: src/ui/reducers/network.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ export type NetworkState = {
   requestBodies: Record<string, RequestBodyData[]>;
   requests: RequestInfo[];
   selectedRequestId: string | null;
+  isCopyTriggered: boolean;
```
Comment (bvaughn) [inline src/ui/reducers/network.ts:25]: Doesn't look like anything is reading or updating this piece of reducer state?

File: src/ui/components/NetworkMonitor/HttpBody.tsx [hunk 3]
```diff
@@ -89,6 +89,7 @@ const HttpBody = ({
       <ReactJson
         style={{ backgroundColor: "none" }}
         theme={theme == "light" ? "rjv-default" : "tube"}
+        name={false}
```
Comment (bvaughn) [inline src/ui/components/NetworkMonitor/HttpBody.tsx:92]: Why is this change necessary/reated?

File: src/ui/components/NetworkMonitor/useCopyToClipboard.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+          `curl '${url}' ${formatRequestHeadersForCURL(requestHeaders)}${requestBodyForCURL}`
+        );
+
+        isCopied = true;
+        shouldShowLoading = false;
+        setTimeout(() => setIsCopyTriggered(false), 3000);
+
+      } else {
+        //dispatch to fetch request body
+        dispatch(fetchRequestBody(id));
```
Comment (bvaughn) [inline src/ui/components/NetworkMonitor/useCopyToClipboard.tsx:71]: Should you be `await`ing this?

File: src/ui/components/NetworkMonitor/RequestTable.tsx (hunk 1 above)
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/RequestTable.tsx]: Ah ! I missed removing the old changes. Thanks for spotting

File: src/ui/reducers/network.ts (hunk 2 above)
Comment (vikneshwar) [reply src/ui/reducers/network.ts]: Same reason as above. Missed removing the old changes

File: src/ui/components/NetworkMonitor/HttpBody.tsx (hunk 3 above)
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/HttpBody.tsx:92]: Not setting the props results in request body json wrapped inside a `root` key, I guess that is not expected. Request Body json should show up exactly what was in the request. Correct me if I am wrong. Thanks

File: src/ui/components/NetworkMonitor/useCopyToClipboard.tsx (hunk 4 above)
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/useCopyToClipboard.tsx]: Moved it to `useEffect`

**Comment by vikneshwar** [comment]
Hi @bvaughn addressed the review comments. Thanks

File: src/ui/components/NetworkMonitor/useNetworkContextMenu.tsx [hunk 5]
```diff
@@ -1,15 +1,30 @@
...
-export default function useNetworkContextMenu(row: Row<RequestSummary>) {
+export default function useNetworkContextMenu({
+  row,
+  onClipboardCopy,
+}: {
+  row: Row<RequestSummary>;
+  onClipboardCopy: () => void;
+}) {
   const dispatch = useAppDispatch();
+  const requestBodies = useAppSelector(getRequestBodies);
```
Comment (bvaughn) [inline src/ui/components/NetworkMonitor/useNetworkContextMenu.tsx:27]: Looks like this new added value is not being used by anything.

File: src/ui/components/NetworkMonitor/useCopyToClipboard.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  useEffect(() => {
+    const fetchBody = async () => {
+      const shouldGetBody = isCopyTriggered && hasRequestBody && !requestBody;
+
+      if (shouldGetBody) {
+        await dispatch(fetchRequestBody(id));
+      }
+    };
+    fetchBody();
+  }, [dispatch, hasRequestBody, id, isCopyTriggered, requestBody]);
```
[the remark this answers is not in the document]
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/useCopyToClipboard.tsx]: I just realized we don't need to `await` for the results here. This is a socket call similar to `selectAndFetchRequest`, the response from the socket is updated to redux store through `NEW_REQUEST_BODY_PARTS `

File: src/ui/components/NetworkMonitor/useCopyToClipboard.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        isCopied = true;
+        shouldShowLoading = false;
+        setTimeout(() => setIsCopyTriggered(false), 3000);
+      } else {
+        isCopied = false;
+        shouldShowLoading = true;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/useCopyToClipboard.tsx]: Have moved it to the effects and also updated the hook to manage flags through state

File: src/ui/components/NetworkMonitor/HeaderGroups.tsx [hunk 8]
```diff
@@ -32,7 +26,7 @@ export function HeaderGroups({
               const { key, ...columnProps } = column.getHeaderProps();
               const isLastColumn = index === headerGroup.headers.length - 1;
               return (
-                <div className={classNames("p-1", styles[column.id])} {...columnProps} key={key}>
+                <div key={key} className={classNames("p-1", styles[column.id])} {...columnProps}>
```
Comment (vikneshwar) [inline src/ui/components/NetworkMonitor/HeaderGroups.tsx:29]: FIxes `key should be unique` warning showing up in the console. Not sure how having the props in the beginning works though

File: src/ui/components/NetworkMonitor/RequestRow.tsx [hunk 9]
```diff
@@ -56,7 +59,7 @@ export const RequestRow = ({
         tabIndex={0}
         data-testid="NetworkMonitor-RequestTable-RequestRow"
       >
-        <div {...row.getRowProps()}>
+        <div {...rowProps}>
```
Comment (vikneshwar) [inline src/ui/components/NetworkMonitor/RequestRow.tsx:62]: Fixes warning about passing the key to div

File: src/ui/components/NetworkMonitor/HeaderGroups.tsx (hunk 8 above)
Comment (bvaughn) [reply src/ui/components/NetworkMonitor/HeaderGroups.tsx:29]: Doesn't look like this should have changed anything, since `columnProps` (the value being spread) had `key` pulled out of it explicitly:
```js
const { key, ...columnProps } = column.getHeaderProps();
```

File: src/ui/components/NetworkMonitor/RequestRow.tsx (hunk 9 above)
Comment (bvaughn) [reply src/ui/components/NetworkMonitor/RequestRow.tsx:62]: Doesn't look like this should have changed anything either. Can you explain why you think it fixed a warning?

**Review by bvaughn — COMMENTED** [review]
Thanks for making the requested changes!

I have a few subjective follow up changed I'd like to make, but I'm going to go ahead and merge this PR and then make those changes myself on top of it 😄

File: src/ui/components/NetworkMonitor/RequestRow.tsx (hunk 9 above)
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/RequestRow.tsx:62]: Hi @bvaughn, `getRowProps` return value contains the `key` so spreading it here in div causes a warning

File: src/ui/components/NetworkMonitor/HeaderGroups.tsx (hunk 8 above)
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/HeaderGroups.tsx:29]: Yeah it does have the `key` pulled out. The warning here I saw is about key not being unique(but they are actually unique, I confirmed that), not sure how putting it in the beginning fixes it. I don't have a solid answer here, but I tried to move it to the beginning and it did work.

File: src/ui/components/NetworkMonitor/RequestTable.tsx [hunk 10]
```diff
@@ -51,6 +53,17 @@ const RequestTable = ({
         className={classNames("flex h-full w-full flex-col overflow-x-auto", styles.request)}
         {...getTableProps()}
       >
+        {isCopied || shouldShowLoading ? (
+          <div className="absolute z-50 grid h-56 grid-cols-1 content-end place-self-center">
+            <div
+              id="showCopied"
+              className={`mb-1.5 flex shrink rounded-lg bg-black p-1.5 text-center text-white opacity-100 shadow-2xl transition-all duration-700 ease-in-out`}
+            >
+              {`${shouldShowLoading ? "Copying" : "Copied"} to clipboard`}
```
[the remark this answers is not in the document]
Comment (vikneshwar) [reply src/ui/components/NetworkMonitor/RequestTable.tsx:62]: 1. This was not due to `shouldShowLoading` flag, but due to `setTimeout` I have used to defer setting the state by `300ms`. I did this so when the tooltip changes from `Copying to clipboard` to `Copied to clipboard the tooltip is not shown for few milliseconds and that gives a clear visual for the user instead of just text changing from `Copying` to `Copied` . Here `Copying` to `Copied` happens only for the case where there is a request body, for all other cases it just displays `Copied` in the first place and yeah that delay with `setTimeout` has caused issue with items where there is no body. 

2. I was not sure how it is generally shown across the application, so I decided to have it centred.
